File: day20.py
#!/usr/bin/env pipenv run python

import collections

from get_input import get_input

class Point(collections.namedtuple("Point", ["x", "y"])):
    def __repr__(self):
        return f"({self.x}, {self.y})"

    def __add__(self, other):
        if not isinstance(other, Point):
            return NotImplemented
        return type(self)(self.x + other.x, self.y + other.y)

Point.DIRECTIONS = {
    "N": Point(0, 1),
    "S": Point(0, -1),
    "E": Point(1, 0),
    "W": Point(-1, 0)
}

def map_distances(room_map):
    queue = [(Point(0, 0), 0)]
    door_counts = {}
    while queue:
        position, doors = queue.pop()
        if position in door_counts:
            continue
        door_counts[position] = doors
        for next_position in room_map[position]:
            queue.append((next_position, doors + 1))
    return door_counts

def part1(room_map):
    """Solution to part 1"""
    return max(map_distances(room_map).values())

def part2(room_map):
    """Solution to part 2"""
    return sum(1 for d in map_distances(room_map).values() if d >= 1000)

def parse(chars):
    chars = chars.strip('$^')
    stack = []
    location = Point(0, 0)
    door_map = collections.defaultdict(set)
    while chars:
        char, chars = chars[0], chars[1:]
        if char in Point.DIRECTIONS:
            new_location = location + Point.DIRECTIONS[char]
            door_map[new_location].add(location)
            door_map[location].add(new_location)
            location = new_location
        elif char == '(':
            stack.append(location)
        elif char == ')':
            location = stack.pop()
        elif char == '|':
            location = stack[-1]
        else:
            raise ValueError(f"What is this {char}")
    if stack != []:
        raise RuntimeError("Not a valid REGEX, paenthesis do not match")
    return door_map

if __name__ == '__main__':
    assert part1(parse("^WNE$")) == 3
    assert part1(parse("^ENWWW(NEEE|SSE(EE|N))$")) == 10
    assert part1(parse("^ENNWSWW(NEWS|)SSSEEN(WNSE|)EE(SWEN|)NNN$")) == 18
    mapping = parse(get_input(day=20, year=2018).strip())
    print("Part 1: {}".format(part1(mapping)))
    print("Part 2: {}".format(part2(mapping)))

File: day20.rs
use std::collections::{HashSet, HashMap, VecDeque};
use std::iter::FromIterator;

use common::get_input;
use common::point::Point;

type Input = HashMap<Point, HashSet<Point>>;

fn map_distances(map: &Input) -> HashMap<Point, usize> {
    let mut queue = VecDeque::from_iter(vec![(Point::new(0, 0), 0)]);
    let mut door_counts: HashMap<Point, usize> = HashMap::new();
    while let Some((position, doors)) = queue.pop_front() {
        if door_counts.contains_key(&position) {
            continue
        }
        door_counts.insert(position, doors);
        for next_position in map.get(&position).unwrap().iter() {
            queue.push_front((next_position.clone(), doors + 1));
        }
    }
    door_counts
}

fn part1(input: &Input) -> usize {
    map_distances(input).values().max().unwrap().clone()
}

fn part2(input: &Input) -> usize {
    map_distances(input).values().filter(|&&v| v >= 1000).count()
}

fn parse(text: String) -> Input {
    let mut location = Point::new(0, 0);
    let mut door_map: Input = HashMap::new();
    let mut stack = VecDeque::new();
    let mut queue = VecDeque::from_iter(text.trim().chars());
    assert!(queue.pop_front() == Some('^'));
    assert!(queue.pop_back() == Some('$'));
    while let Some(dir) = queue.pop_front() {
        match dir {
            '(' => { stack.push_front(location); },
            ')' => { location = stack.pop_front().unwrap(); },
            '|' => { location = stack.front().unwrap().clone(); },
            _ => {
                let next_location = location + match dir {
                    'N' => Point::new(0, 1),
                    'S' => Point::new(0, -1),
                    'E' => Point::new(1, 0),
                    'W' => Point::new(-1, 0),
                    c => panic!("What is this? {}", c),
                };
                door_map.entry(next_location)
                    .or_insert(HashSet::new())
                    .insert(location);
                door_map.entry(location)
                    .or_insert(HashSet::new())
                    .insert(next_location);
                location = next_location;
            },
        }
    }
    assert!(stack.len() == 0);
    door_map
}

fn main() {
    assert!(part1(&parse("^WNE$".to_string())) == 3);
    let input = parse(get_input(20, 2018));
    println!("Part 1: {}", part1(&input));
    println!("Part 2: {}", part2(&input));
}
